internal/nexodus/keys.go

```diff
@@ -12,18 +14,112 @@ import (
-	darwinPublicKeyFile   = "/usr/local/etc/wireguard/public.key"
-	darwinPrivateKeyFile  = "/usr/local/etc/wireguard/private.key"
-	windowsPublicKeyFile  = "C:/nexd/public.key"
-	windowsPrivateKeyFile = "C:/nexd/private.key"
 	publicKeyPermissions  = 0644
 	privateKeyPermissions = 0600
 )
 
+// handleKeys will look for an existing key pair, if a pair is not found this method
+// will generate a new pair and write them to location on the disk depending on the OS
+func (ax *Nexodus) handleKeys() error {
+
+	pubKeyFile := filepath.Join(ax.stateDir, "public.key")
+	privKeyFile := filepath.Join(ax.stateDir, "private.key")
+
+	// We used to store the keys in a different location..
+	// migrate them to the state dir if needed.
+	err := ax.migrationCheck(pubKeyFile, privKeyFile)
+	if err != nil {
+		return err
+	}
+
+	publicKey := readKeyFile(ax.logger, pubKeyFile)
+	privateKey := readKeyFile(ax.logger, privKeyFile)
+	if publicKey != "" && privateKey != "" {
+		ax.wireguardPubKey = publicKey
+		ax.wireguardPvtKey = privateKey
+		ax.logger.Infof("Existing key pair found at [ %s ] and [ %s ]", pubKeyFile, privKeyFile)
+		return nil
+	}
+	ax.logger.Infof("No existing public/private key pair found, generating a new pair")
+	if err := ax.generateKeyPair(pubKeyFile, privKeyFile); err != nil {
+		return fmt.Errorf("Unable to locate or generate a key/pair: %w", err)
+	}
+	ax.logger.Debugf("New keys were written to [ %s ] and [ %s ]", pubKeyFile, privKeyFile)
+	return nil
+}
+
+// migrationCheck should not be needed after everyone has upgraded to the latest nexd.
+func (ax *Nexodus) migrationCheck(pubKeyFile string, privKeyFile string) error {
+
+	// skip if the new key files exist
+	if canReadFile(pubKeyFile) && canReadFile(privKeyFile) {
+		return nil
+	}
+
+	oldPubKeyFile := "public.key"
+	oldPrivKeyFile := "private.key"
+	if !ax.userspaceMode {
+		switch runtime.GOOS {
+		case "darwin":
+			oldPubKeyFile = "/usr/local/etc/wireguard/public.key"
+			oldPrivKeyFile = "/usr/local/etc/wireguard/private.key"
+		case "windows":
+			oldPubKeyFile = "C:/nexd/public.key"
+			oldPrivKeyFile = "C:/nexd/private.key"
+		case "linux":
+			oldPubKeyFile = "/etc/wireguard/public.key"
+			oldPrivKeyFile = "/etc/wireguard/private.key"
+		}
+	}
+
+	// skip if they being stored in the same place.
+	if pubKeyFile == oldPubKeyFile || privKeyFile == oldPrivKeyFile {
+		return nil
+	}
+
+	// skip if the old key files don't exist
+	if !(canReadFile(oldPubKeyFile) && canReadFile(oldPrivKeyFile)) {
+		return nil
+	}
+
+	data, err := os.ReadFile(oldPubKeyFile)
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(pubKeyFile, data, publicKeyPermissions)
+	if err != nil {
+		return err
+	}
+	data, err = os.ReadFile(oldPrivKeyFile)
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(privKeyFile, data, privateKeyPermissions)
+	if err != nil {
+		return err
+	}
+
+	// TODO: decide if we should delete the old keys.
+	//err = os.Remove(oldPubKeyFile)
```

Comment:
@russellb do you think we should delete the old keys?

Reply:
No, because we don't know where the old keys came from. If we didn't generate them, we need to leave them in place. A user may have other uses of wireguard on the system.

I wonder if it should be a preference order kind of thing.

1. Use `/etc/wireguard` if keys are there.
2. Use `statedir/` if keys are there.
3. Generate new keys and place in `statedir/`

Reply:
I think we should NOT look at /etc/wireguard at all at some point the future since that is likely to be user managed WG interface configuration.  I see this copy step as a temporary migration path to ensure smooth upgrade to our current users.

Reply:
I think we should NOT look at /etc/wireguard at all at some point the future since that is likely to be user managed WG interface configuration.  I see this copy step as a temporary migration path to ensure smooth upgrade to our current users.

Reply:
OK, how about:

1. Use statedir keys.
2. If /etc/wireguard has keys AND the public key matches what's already configured on the device, copy it / use it.
3. Generate new keys in statedir.

Reply:
and actually, I'm not sure statedir (/var/lib/) is the right place for keys. I think it's /etc. We could use /etc/nexodus instead of /etc/wireguard, though.

For reference: https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html

> /var contains variable data files. This includes spool directories and files, administrative and logging data, and transient and temporary files.

and for /var/lib

> This hierarchy holds state information pertaining to an application or the system. State information is data that programs modify while they run, and that pertains to one specific host. Users must never need to modify files in /var/lib to configure a package's operation, and the specific file hierarchy used to store the data must not be exposed to regular users. [[40]](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html#ftn.idm236090988912)

Reply:
and this page has some commentary about storing user-specific configuration, like the keys when running in non-privileged mode:

https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s08.html#homeReferences

It suggests `~/.nexodus/...`. If we have our own dir, we could use it for both state (cached tokens) and config (keys).

Reply:
I see device keys and user auth tokens to be about the same.  It's all nexd runtime state to me.  Given we generate the keys and the tokens, you could argue it's not configuration so it should NOT live in /etc.

I'm kind of partial to putting all that stuff in:
* if user == root: /var/lib/nexodus
* if user != root: ~/.nexodus

Reply:
The keys are associated with device identity, while the token is session-related. We may or may not create the keys. We will always create the cached tokens.

Reply:
It would be simpler if we just simply did not support the user creating the keys.  In fact that could introduce more failure conditions.  What if he reuses the same keys across different devices?

Reply:
They would have a broken setup if a key was reused. I imagine the two `nexd` instances would fight over resetting the values on the same device.

I don't know how to tell the difference between keys we generated vs. ones placed on disk for us, though, so that'll always be a possibility.